### PR TITLE
Add Template API endpoint

### PR DESCRIPTION
**Pull request by Rashmil-1999**

### CODE REVIEW

**Comment by Rashmil-1999** [comment]
In this PR I add a new endpoint `semantics/api/types/<name>/template`
To test it:
- make a `Send and Download` request in Postman using any current valid dataset type present in prod. Right now the only way to determine shapefile from CSV is the field `geom`, `geometry`, and `the_geom`. It may or may not be present, so I recommend testing it, either changing the boolean shapefile variable to true on a CSV type datatype or finding a datatype with this param.
- Either way, it will generate CSV format even for shapefile datatypes as there is no geometry column name present in the majority of the shapefile-based datatypes.

Other changes:

- Added a utility method to fetch typelist for 3 endpoints removing code duplication.

**Comment by ywkim312** [comment]
@Rashmil-1999 Can you add some example end point to test for shapefiles or csvs? I already know csv though =) but for other reviewers?

**Comment by ywkim312** [comment]
> In this PR I add a new endpoint `semantics/api/types/<name>/template` To test it:
> 
> * make a `Send and Download` request in Postman using any current valid dataset type present in prod. Right now the only way to determine shapefile from CSV is the field `geom`, `geometry`, and `the_geom`. It may or may not be present, so I recommend testing it, either changing the boolean shapefile variable to true on a CSV type datatype or finding a datatype with this param.
> * Either way, it will generate CSV format even for shapefile datatypes as there is no geometry column name present in the majority of the shapefile-based datatypes.
> 
> Other changes:
> 
> * Added a utility method to fetch typelist for 3 endpoints removing code duplication.

I am not very sure if the endpoint create csv for the shapefile? Or create shapefile for shapefile type?

**Comment by ylyangtw** [comment]
I don't see our favorite friend - changelog

**Comment by ylyangtw** [comment]
I am getting this... is it ok? 
image

**Comment by Rashmil-1999** [comment]
> > In this PR I add a new endpoint `semantics/api/types/<name>/template` To test it:
> > 
> > * make a `Send and Download` request in Postman using any current valid dataset type present in prod. Right now the only way to determine shapefile from CSV is the field `geom`, `geometry`, and `the_geom`. It may or may not be present, so I recommend testing it, either changing the boolean shapefile variable to true on a CSV type datatype or finding a datatype with this param.
> > * Either way, it will generate CSV format even for shapefile datatypes as there is no geometry column name present in the majority of the shapefile-based datatypes.
> > 
> > Other changes:
> > 
> > * Added a utility method to fetch typelist for 3 endpoints removing code duplication.
> 
> I am not very sure if the endpoint create csv for the shapefile? Or create shapefile for shapefile type?

So this endpoint will return a shapefile if the dataset type has `geom` | `the_geom` | `geometry` column otherwise it will return a CSV. This is under the assumption that we only have datatypes defined for table and shapefile format datasets as there is no obvious way to determine what format the dataset is in.

**Comment by Rashmil-1999** [comment]
> I don't see our favorite friend - changelog:P

Ahh.... Changelog is sneaky, it always tend to slip from my mind hahahah

**Comment by Rashmil-1999** [comment]
Screenshot 2023-09-26 at 3 40 24 PM

@ylyangtw  try the `Send and Download` option.

File: server/semantics-service/src/main/java/edu/illinois/ncsa/incore/service/semantics/utils/GeotoolsUtils.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import org.geotools.feature.DefaultFeatureCollection;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
+import org.geotools.referencing.crs.DefaultGeographicCRS;
+import org.locationtech.jts.geom.Point;
+import org.opengis.feature.simple.SimpleFeatureType;
+
+import java.io.*;
+import java.util.*;
+
+public class GeotoolsUtils {
```
Comment (longshuicy) [inline server/semantics-service/src/main/java/edu/illinois/ncsa/incore/service/semantics/utils/GeotoolsUtils.java:19]: Do you just decide to copy the same GeotoolsUtils to semantics service?
I think im ok with that but you might want to create subsequent task to put this and the fileUtil to incore-common

**Review by longshuicy — CHANGES_REQUESTED** [review]
Can you add a changelog entry? Code looks good otherwise

**Comment by ylyangtw** [comment]
Tried ergo_bridgeDamage and it works. Approve~
image

**Review by ylyangtw — APPROVED** [review]

File: server/semantics-service/src/main/java/edu/illinois/ncsa/incore/service/semantics/utils/GeotoolsUtils.java (hunk 1 above)
Comment (Rashmil-1999) [reply server/semantics-service/src/main/java/edu/illinois/ncsa/incore/service/semantics/utils/GeotoolsUtils.java:19]: Yes, I tried doing that but some of these fileutils and geotoolutils methods would stop working if I changed directories as it required some stuff from other files. Thats why I thought to just copy those functions to semantics

Comment (longshuicy) [reply server/semantics-service/src/main/java/edu/illinois/ncsa/incore/service/semantics/utils/GeotoolsUtils.java:19]: I can create a task to document that.

**Review by longshuicy — APPROVED** [review]

**Pull request merged**